File: HackerRank/saveprincess.py
#!/bin/python
# Head ends here
MOVES = ['LEFT','RIGHT','UP','DOWN']
def displayPathtoPrincess(M,grid):
    m,p = (i for i,x in enumerate(grid) if x != '-')
    if grid[m] == 'p': m,p = p,m
    m,p = ((x/M, x%M) for x in [m,p])
    if m[0] > p[0]:
        v = 2
    else:
        v = 3
    if m[1] > p[1]:
        h = 0
    else:
        h = 1

    path = [v] * abs(m[0] - p[0])
    path.extend([h] * abs(m[1] - p[1]))
    return '\n'.join([MOVES[p] for p in path])
# Tail starts here
m = int(input())

#grid = []
#for i in xrange(0, m):
    #grid.append(raw_input().strip())
grid = ''.join([raw_input().strip() for _ in xrange(m)])

print displayPathtoPrincess(m,grid)

File: HackerRank/saveprincess.scala
object Solution {
  def main(args: Array[String]) = { 
    val m :Int = Integer.parseInt(Console.readLine)
    val grid = (0 until m).map( x => Console.readLine ) mkString ""
    displayPathtoPrincess(m,grid);
  }
  def displayPathtoPrincess(M:Int,grid:String)={
    val found = grid.zipWithIndex.filter(x => x._1 != '-')
    val List(p,m) = ((if (found(0)._1 == 'p') found else found.reverse).toList).map(x => (x._2 % M, x._2 / M))
    val h = if (p._1>m._1) "RIGHT" else "LEFT"
    val v = if (p._2>m._2) "UP" else "DOWN"
    for( _ <- 0 until (p._1-m._1).abs ) println(h)
    for( _ <- 0 until (p._2-m._2).abs ) println(v)
  }
}
